pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -241,13 +241,19 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceConfig rebala
           tierToInstancePartitionsMap, null);
     }
 
+    // Record the beginning of rebalance
+    _tableRebalanceObserver.onTrigger(TableRebalanceObserver.Trigger.START_TRIGGER, currentAssignment,
+        targetAssignment);
+
     boolean segmentAssignmentUnchanged = currentAssignment.equals(targetAssignment);
     LOGGER.info("For rebalanceId: {}, instancePartitionsUnchanged: {}, tierInstancePartitionsUnchanged: {}, "
             + "segmentAssignmentUnchanged: {} for table: {}", rebalanceJobId, instancePartitionsUnchanged,
         tierInstancePartitionsUnchanged, segmentAssignmentUnchanged, tableNameWithType);
 
     if (segmentAssignmentUnchanged) {
-      LOGGER.info("Table: {} is already balanced", tableNameWithType);
+      String msg = String.format("Table: %s is already balanced", tableNameWithType);
+      LOGGER.info(msg);
+      _tableRebalanceObserver.onSuccess(msg);
```

Comment:
> If you rebalance a table with "reassignAssignments" and the instance assignment doesn't change, you will get "IN_PROGRESS" back, but the id won't exist in zk 

I'd assume we got a DONE status at L255 `if (dryRun) {...}` firstly (because rebalance restful API always does a dryRun firstly and triggers real trigger if dryRun gives DONE). But likely, the real table rebalance must have ended at the L259 `else {...}`. Because the observer.onTrigger(...START_TRIGGER) was run much later, so no job status got put in ZK for the jobId.

If that's the case, then I'm wondering why we didn't return NO_NP for all three if branches below? cc @Jackie-Jiang

Reply:
and in general, I'd prefer we just track job status in `ZK` for table rebalances that really move segments around and potentially take time to finish, so returning NO_OP for all cases of 'table is already balanced'